clang/include/clang/Sema/Sema.h

```diff
@@ -12963,10 +12963,26 @@ void Sema::addIntelSYCLSingleArgFunctionAttr(Decl *D,
       return;
     }
     int32_t ArgInt = ArgVal->getSExtValue();
-    if (ArgInt <= 0) {
-      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
-          << CI.getAttrName() << /*positive*/ 0;
-      return;
+    if (CI.getParsedKind() == ParsedAttr::AT_SYCLIntelNumSimdWorkItems ||
+        CI.getParsedKind() == ParsedAttr::AT_IntelReqdSubGroupSize) {
+      if (ArgInt <= 0) {
+        Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+            << CI.getAttrName() << /*non-negative*/ 1;
+        return;
+      }
+    }
+    if (CI.getParsedKind() == ParsedAttr::AT_SYCLIntelMaxGlobalWorkDim) {
+      if (ArgInt < 0) {
+        Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+            << CI.getAttrName() << /*positive*/ 0;
+        return;
+      }
+      if (ArgInt > 3) {
+        Diag(E->getExprLoc(),
+             diag::err_intel_attribute_argument_is_not_in_range)
```

Comment:
I think that the diagnostic message provided by `err_argument_invalid_range` should be clear enough to avoid adding `err_intel_attribute_argument_is_not_in_range`.
Here are both of them to compare.
- `err_intel_attribute_argument_is_not_in_range`: "The value of %0 attribute must be in range from 0 to 3">;
- `err_argument_invalid_range`: "argument value %0 is outside the valid range [0, 3]"

Thoughts?

Reply:
Actually after I replaced %0 with actual attribute name, I can see that `err_argument_invalid_range` refers to `max_global_work_dim` as argument name, but it's an attribute.

Reply:
@bader thanks for looking into the patch. I tried new diagnostic you suggested but it is causing assertion like this: 
Assertion `Idx < getNumArgs() && "Argument index out of range!"' failed.

///  It looks like we are using this diagnostic "err_argument_invalid_range" to 
/// handle a check if argument ArgNum of CallExpr
/// TheCall is a constant expression in the range [Low, High].

bool Sema::SemaBuiltinConstantArgRange(CallExpr *TheCall, int ArgNum,
                                       int Low, int High, bool RangeIsError) {
  


      return Diag(TheCall->getBeginLoc(), diag::err_argument_invalid_range)

Reply:
Before fixing the assertion, I'd like someone who speaks English natively confirm that diagnostics provided by `err_argument_invalid_range` is not confusing. 
Otherwise, I'm okay with adding new intel specific diagnostic and we can consider this resolved. 

With this comment, I'm trying to minimize the deviation from clean llvm-project source, so it's not that critical.

Reply:
@bader, I understand your point here. 

The diagnostics provided by err_argument_invalid_range is not confusing. 

While working on fixing the assertion, I noticed that we have similar diagnostic “err_attribute_argument_out_of_range” that seems more applicable to the attribute here.

I have updated current patch with it. If you have preference to other, I can change it to “err_argument_invalid_range” 

err_attribute_argument_out_of_range: 'max_global_work_dim' attribute requires integer constant between 0 and 3 inclusive
err_argument_invalid_range: argument value 'max_global_work_dim' is outside the valid range [0, 3]

Reply:
`err_attribute_argument_out_of_range` seems to be exactly what is needed here. Thanks!